compiler/rustc_mir_transform/src/log_gc_stats.rs

```diff
@@ -0,0 +1,90 @@
+#![cfg_attr(not(feature = "rustc_log_gc_stats"), allow(dead_code))]
+use std::io::Write;
+use std::{env, fs};
+
+use rustc_hir::def_id::DefId;
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+use rustc_span::sym;
+use tracing::trace;
+
+use crate::MirPass;
+
+#[derive(Default)]
+struct GcStats {
+    num_gcs: u64,
+    num_rcs: u64,
+    num_weaks: u64,
+    num_arcs: u64,
+    num_arcweaks: u64,
+}
+
+pub(super) struct LogGcStats;
+
+impl<'tcx> MirPass<'tcx> for LogGcStats {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        if env::var("ALLOY_RUSTC_LOG").is_err() {
+            return;
+        }
+        trace!("Calculating GcStats on {:?}", body.source);
+
+        if in_std_lib(tcx, body.source.def_id()) {
+            // A hacky way of checking if we're in the standard library.
+            // If we are, we don't want to record GC statistics.
+            return;
+        }
+
+        let gc = tcx.get_diagnostic_item(sym::gc).unwrap();
+        let rc = tcx.get_diagnostic_item(sym::Rc).unwrap();
+        let arc = tcx.get_diagnostic_item(sym::Arc).unwrap();
+        let weak = tcx.get_diagnostic_item(sym::RcWeak).unwrap();
+        let arcweak = tcx.get_diagnostic_item(sym::ArcWeak).unwrap();
+
+        let mut stats = GcStats::default();
+
+        for decl in body.local_decls().iter().skip(1) {
+            if decl.ty.ty_adt_def().is_none() {
+                continue;
+            }
+
+            let did = decl.ty.ty_adt_def().unwrap().did();
+            match did {
+                _ if did == gc => stats.num_gcs += 1,
+                _ if did == rc => stats.num_rcs += 1,
+                _ if did == arc => stats.num_arcs += 1,
+                _ if did == weak => stats.num_weaks += 1,
+                _ if did == arcweak => stats.num_arcweaks += 1,
+                _ => (),
+            }
+        }
+
+        let mut filename = fs::OpenOptions::new()
+            .write(true)
+            .create(true)
+            .append(true)
+            .open(env::var("ALLOY_RUSTC_LOG").unwrap())
```

Comment:
Do we want to handle `-` as meaning "output to stdout/stderr" as a special case here?

Reply:
The difficulty with this is that this streams the filewriting as it compiles each function (i.e. appending to the same file as it goes along). So you'd end up with stdout/stderr output intertwined with the regular cargo `Compiling ...` output.

Reply:
OK.